### PR TITLE
[flang] Review "auto" usage in Lower/Allocatable.cpp

### PR DESCRIPTION
**Pull request by klausler**
Audit the use of C++'s "auto" declarations in Allocatable.cpp
and replace some instances of "auto" with explicit type names.
Remaining uses conform with LLVM's style guide.

### CODE REVIEW

**Review by clementval — APPROVED** [review]
LGTM. Thanks.

File: flang/lib/Lower/Allocatable.cpp [hunk 1]
```diff
@@ -55,7 +55,7 @@ struct ErrorManager {
             const Fortran::lower::SomeExpr *statExpr,
             const Fortran::lower::SomeExpr *errMsgExpr) {
     Fortran::lower::StatementContext stmtCtx;
-    auto &builder = converter.getFirOpBuilder();
+    fir::FirOpBuilder &builder = converter.getFirOpBuilder();
```
Comment (schweitzpgi) [inline flang/lib/Lower/Allocatable.cpp:58]: Straddles the fuzzy line a bit. The `getFoo` is sometimes enough, sometimes not, depending on the reviewer.

File: flang/lib/Lower/Allocatable.cpp [hunk 2]
```diff
@@ -137,25 +138,25 @@ static void genRuntimeInitCharacter(fir::FirOpBuilder &builder,
...
-  auto inputTypes = callee.getType().getInputs();
+  llvm::ArrayRef<mlir::Type> inputTypes = callee.getType().getInputs();
   if (inputTypes.size() != 5)
     fir::emitFatalError(
         loc, "AllocatableInitCharacter runtime interface not as expected");
   llvm::SmallVector<mlir::Value> args;
   args.push_back(builder.createConvert(loc, inputTypes[0], box.getAddr()));
   args.push_back(builder.createConvert(loc, inputTypes[1], len));
-  auto kind = box.getEleTy().cast<fir::CharacterType>().getFKind();
+  int kind = box.getEleTy().cast<fir::CharacterType>().getFKind();
```
Comment (schweitzpgi) [inline flang/lib/Lower/Allocatable.cpp:154]: This is a silent type conversion. getFKind returns a `KindTy`. And a `KindTy` is actually `unsigned int`.

Probably stick with `Fortran::lower::KindTy`.

File: flang/lib/Lower/Allocatable.cpp [hunk 3]
```diff
@@ -518,7 +530,7 @@ void Fortran::lower::genDeallocateStmt(
     const Fortran::parser::DeallocateStmt &stmt, mlir::Location loc) {
   const Fortran::lower::SomeExpr *statExpr{nullptr};
   const Fortran::lower::SomeExpr *errMsgExpr{nullptr};
-  for (const auto &statOrErr :
+  for (const Fortran::parser::StatOrErrmsg &statOrErr :
        std::get<std::list<Fortran::parser::StatOrErrmsg>>(stmt.t))
```
Comment (schweitzpgi) [inline flang/lib/Lower/Allocatable.cpp:534]: Type is clear from the `std::get`, no?

File: flang/lib/Lower/Allocatable.cpp [hunk 4]
```diff
@@ -556,15 +569,15 @@ isNonContiguousArrayPointer(const Fortran::semantics::Symbol &sym) {
...
+  Fortran::semantics::Scope::Kind kind = owner.kind();
   // Test if this is a procedure scope that contains a subprogram scope that is
   // not an interface.
   if (kind == Fortran::semantics::Scope::Kind::Subprogram ||
       kind == Fortran::semantics::Scope::Kind::MainProgram)
-    for (const auto &childScope : owner.children())
+    for (const Fortran::semantics::Scope &childScope : owner.children())
       if (childScope.kind() == Fortran::semantics::Scope::Kind::Subprogram)
-        if (const auto *childSym = childScope.symbol())
+        if (const Fortran::semantics::Symbol *childSym = childScope.symbol())
```
Comment (schweitzpgi) [inline flang/lib/Lower/Allocatable.cpp:580]: I guess this one doesn't start with `get` so might be confusing. :)

File: flang/lib/Lower/Allocatable.cpp [hunk 5]
```diff
@@ -531,11 +543,12 @@ void Fortran::lower::genDeallocateStmt(
                statOrErr.u);
   ErrorManager errorManager;
   errorManager.init(converter, loc, statExpr, errMsgExpr);
-  auto &builder = converter.getFirOpBuilder();
-  auto insertPt = builder.saveInsertionPoint();
-  for (const auto &allocateObject :
+  fir::FirOpBuilder &builder = converter.getFirOpBuilder();
+  mlir::OpBuilder::InsertPoint insertPt = builder.saveInsertionPoint();
+  for (const Fortran::parser::AllocateObject &allocateObject :
        std::get<std::list<Fortran::parser::AllocateObject>>(stmt.t)) {
```
Comment (schweitzpgi) [inline flang/lib/Lower/Allocatable.cpp:549]: This block seems to echo types on these 3 lines to me.

**Review by schweitzpgi — APPROVED** [review]
Thanks, Peter.

File: flang/lib/Lower/Allocatable.cpp (hunk 1 above)
Comment (klausler) [reply flang/lib/Lower/Allocatable.cpp:58]: Does Mehdi accept "auto" here?

File: flang/lib/Lower/Allocatable.cpp (hunk 2 above)
Comment (klausler) [reply flang/lib/Lower/Allocatable.cpp:154]: Braced inits would have caught that for me, grr.

File: flang/lib/Lower/Allocatable.cpp (hunk 3 above)
Comment (klausler) [reply flang/lib/Lower/Allocatable.cpp:534]: If you think that's okay, sure.

File: flang/lib/Lower/Allocatable.cpp (hunk 1 above)
Comment (klausler) [reply flang/lib/Lower/Allocatable.cpp:58]: Do you know whether Mehdi would reject `auto` when the initializer is a call to `get<TYPE>()`?

Comment (clementval) [reply flang/lib/Lower/Allocatable.cpp:58]: They accept `auto` in MLIR where they have `auto loc = op.getLoc()` and the actual type should be `mlir::Location`. 

`auto` should be fine when the initializer is a call to `get<TYPE>()`.

File: flang/lib/Lower/Allocatable.cpp (hunk 5 above)
Comment (schweitzpgi) [reply flang/lib/Lower/Allocatable.cpp:549]: If I wasn't clear, it reads:

builder builder get builder;
insertion point insertion point save insertion point;
fortran parser allocate object allocate object get list fortran parser allocate object;

**Pull request merged**